core/src/main/scala/org/bykn/bosatsu/DefRecursionCheck.scala

```diff
@@ -165,11 +184,22 @@ object DefRecursionCheck {
      * Check that decl is a strict substructure of pat. We do this by making sure decl is a Var
      * and that var is one of the strict substrutures of the pattern.
      */
-    def strictSubstructure(fnname: Bindable, pat: Pattern.Parsed, decl: Declaration): Res =
+    private def strictSubstructure(fnname: Bindable, pat: Pattern.Parsed, decl: Declaration, optApplyState: Option[ApplyState]): Res =
       decl match {
-        case v@Declaration.Var(nm) =>
+        case v@Declaration.Var(nm: Bindable) =>
           if (pat.substructures.contains(nm)) unitValid
-          else Validated.invalidNel(RecursionNotSubstructural(fnname, pat, v))
+          else {
+            optApplyState match {
+              case None => Validated.invalidNel(RecursionNotSubstructural(fnname, pat, v))
+              case Some(ApplyState(_, _, apV, binds)) =>
```

Comment:
seems strange that we are ignoring the function name here... can we remove it entirely? What about the parent?